content/blog/2020-06-28-shtab-completion.md

```diff
@@ -0,0 +1,197 @@
+---
+title: '(Tab) Complete Any Python Application in 1 Minute or Less'
+date: 2020-06-28
+description: |
+  We've made a painless tab-completion script generator for Python applications!
+  Find out how to take advantage of it in this blog post.
+
+descriptionLong: |
+  We've made a painless tab-completion script generator for Python applications!
+  It's called `shtab` and it currently works with `argparse`, `docopt`, and
+  `argopt` to produce `bash` and `zsh` completion scripts.
+  This tool was originally created to help `dvc`, but we realised it could be made
+  more generic and valuable to the world's entire ecosystem of Python CLI
+  applications. Find out how to take advantage of it in this blog post.
+author: casper_dcl
+---
+
+> Other potential titles:
+>
+> - Zero Effort Tab Completion for Python Applications
+> - Finally: 1-Click Tab Completion for [`argparse`], [`docopt`] and [`argopt`]
+> - A [`shtab`] in the dark at completing your ~~sentences~~ commands
+
+Command line tools are powerful. Things like [`make`] have manual pages
+spanning, well,
+[pages](https://www.gnu.org/software/make/manual/make.html#Options-Summary),
+while just the list of [`git`] subcommands is longer than can fit on a standard
+`80 x 24` terminal screen.
+
+```dvc
+$ git <TAB>
+add                  filter-branch        rebase
+am                   format-patch         reflog
+annotate             fsck                 relink
+...
+describe             prco                 unassume
+--More--
+```
+
+Notice the `--More--` at the bottom? That's the joy of pagination.
+
+Notice the `<TAB>` at the top? That represents actually pressing the tab key.
+Ah, the joy of shell tab completion.
+
+Tab completion is an indispensable part of writing anything on the command-line.
+Personally, I can't imaging trying to `git co` (aliased to `git checkout`) a
+branch without `<TAB>` to do the heavy lifting.
+
+Now there's a tool called `dvc` which is like [Git for data](https://dvc.org).
+It can be viewed as a cross-platform combination of [`git`] and [`make`]
+designed for handling big data and multiple cloud storage repositories, as well
+as tracking machine learning experiments. As you can imagine, supporting that
+many buzzwords means it also has a large number of subcommands and options.
+
+_Every time a new feature is added, maintainers and contributors have to update
+tab completion scripts for multiple supported shells. At best, it's a pain, and
```

Comment:
Maybe link to the tab completion scripts in an old version of the repo? Or to their commit history

Reply:
good idea. Holy moly:

<details>

```
fa0f182 completion: autogenerate (#3921)
13012f9 plots: completion/spelling fixups (#3952)
9a55b90 Rename --xlab to --x-label, X_TITLE to X_LABEL, same for y (#3948)
ac4f351 dvc: require `--external` for external outputs (#3929)
86f3457 dvc: remove `dvc metrics add/remove` (#3928)
d1e08b6 dvc: replace `pipeline` with `dag` (#3905)
9ee62ef dvc: rename "working tree" to "workspace" (#3914)
0bb49d2 dvc: rename `dvc lock` -> `dvc freeze` (#3915)
c5920b0 Auto-completion of the `dvc remote rename` command (#3868)
865d355 params/metrics: introduce --no-path and --old (#3845)
130b76c plots: remove --select CLI option (#3840)
19eaced dvc : renamed --outs-only to --dvc-only (#3797)
cec4b4e dvc: rename plot to plots (#3802)
974cf09 run/repro: rename "build cache" -> "run cache" (#3788)
cc16b28 diff: autocompletion script for params, and show-md (#3786)
af8ac00 plot: completion for zsh (#3733)
e553511 Metrics - plotting for multiple revisions initial (#3577)
e05157f completion: bash: fix files (#3613)
f325921 diff: update help output to match docs, et al. (#3601)
d2ea180 pull: update help output (-R option) (#3555)
e133444 completion: add --summary flag on checkout (#3491)
dcf5af6 list: rename target param to path, update docs (#3462)
7923446 completion: bash: misc fixes and updates (#3467)
ea981ae gc: make --all-commits flag visible (#3429)
eebbf3f change diff checksum arg name to hash (#3384)
411057d completion: bash: properly recurse targets (#3374)
2820aed completion: bash: allow multi-arg completion, minor tidy (#3367)
a9161af Fix bash completion (#3360)
b23f43a term: avoid "under X control" and other help output updates (#3322)
46e5053 update: support specifying revision when updating (#3337)
7e6e8ee api: add list command (#3246)
de1f67a Merge pull request #3244 from jorgeorpinel/master
4e87bf9 try beautysh
611c1d6 scripts: update autocomplete per diff changes per https://github.com/iterative/dvc/pull/3244#pullrequestreview-351108115
98d2ca0 lint: shfmt with -ci
10c5103 lint: run shfmt -i 2
2e79ce3 lint: run shfmt
aa849ab Merge branch 'master' of github.com:iterative/dvc
d0139a7 term: review "revision", "reference", and "hash", "SHA" terms
7f6324e diff: remove --target
fced171 diff: --json -> --show-json
27ada02 scripts: update completion scripts for `diff`
6ad55ec completion: get --show-url
8712ecc completion: fix --relink problem
bd0b28d ui: add checkout --relink to autocomplete scripts
461a79e Fix typos
ed89725 Replace hyphen with underscore for command auto-completion (#2596)
7059ccb update completion scripts and remove show_checksums from status
2930ea7 remove show-checksums flag for dvc status commmand
80c30ce update: better short desc in command help outut
0575104 update: small autocomplete scripts reformat
c7ef659 output: remove a couple more "(s)" left
6f5449f term: remove "(s)" and "(are) related to the previous commit
d4f4126 help: finish singularizing "pipeline" term...
89541b7 Autocompletion of commands and files (#2226)
db19957 Download refactor (#2275)
932b0d7 introduce `dvc update` for ext repos (#2218)
8adc313 change
bd407a6 changed " to '
47398c0 Resolved conflicts
d440cae resolve conflicts
40617fa Revert "Revert "Revert "change"""
0f540b8 Revert "Revert "change""
b835bbe Revert "change"
829c800 change
b6b77dd removed daemon
55d15ed change
e04e4c0 scripts: add import and get commands (et al) to autocompletes
fd12a0e scripts: update `import`->`import-url` and add `get-url` to shell autocomplete
b5eb4c4 -V for version
9184976 Updated dvc and bash
18711ec more changes to be made
68594bf Updated bash
5b2405f scripts: fix zsh autocomplete for `dvc pipeline` subcommands
5ff8f3b updated bash autocomplete script
3dffe8e dvc: i/DVC file[s]?/ -> /DVC-file($1)/
8df0712 update zsh autocomplete script with current command help messages + (#2064)
ce83231 Updates to 1. zsh autocomplete script and 2. slow link type warning message (#2003)
c0104e2 repro: implement downstream option (#1926)
f417f49 dvc: deprecate --remove-outs
ce65771 scripts: update completion for zsh and bash
19007ff scripts: Add file completion to `dvc add` command (#1423)
9837625 Add draft for DVC shell completion in Bash
31b7a57 Add draft for DVC shell completion in Zsh
```

</details>

Reply:
🔩

Reply:
added